src/adapter/src/coord.rs

```diff
@@ -2630,7 +2630,25 @@ impl Coordinator {
             // Index dataflow.
 
             write_frontier = self.least_valid_write(&storage_dependencies);
-            if let Some(ts) = write_frontier.as_option() {
+
+            if self.controller.read_only() {
+                // Choose something that is most likely to make dependent indexes
+                // hydrate. But this will likely make it so that reconciliation
+                // cannot re-use already-running dataflows.
+                //
+                // The latter is fine, though, because in read-only mode we're
+                // likely bringing up new cluster processes anyways.
+                //
+                // NOTE: This is working around a quirk in how we normally
+                // select as_ofs for indexes: we don't take dependent indexes
+                // into account when bounding the as_of, so it can happen that
+                // we end up in a situation where a downstream index cannot
+                // hydrate even though it's transitive storage collection
+                // dependencies would have the data at the as_of that is
+                // required. This is a band-aid solution until we have
+                // https://github.com/MaterializeInc/materialize/issues/28440
+                min_as_of.clone()
```

Comment:
This sounds good to me. But don't we need it for MV dataflows too? At least `REFRESH EVERY` MVs can have their write frontier far in the future, so we could pick an as-of that doesn't allow them to rehydrate immediately as well I think.

Reply:
And you think exactly the same would work for them as well?

Reply:
Force-pushed a change that adds it for MVs.

Reply:
Using the `min_as_of` should always work! There is some risk that the thing we are reading from has a `since` that's really far in the past and the replica will have to crunch through a lot of uncompacted historical data to catch up, which might cause it to OOM. So it's good that we don't treat this as a long-term fix.